src/pint/pulsar_mjd.py

```diff
@@ -56,25 +59,38 @@ class TimePulsarMJD(TimeFormat):
     name = 'pulsar_mjd'
     # This can be removed once we only support astropy >=3.1.
     # The str(c) is necessary for python2/numpy -> no unicode literals...
-    _new_ihmsfs_dtype = numpy.dtype([(str(c), numpy.intc) for c in 'hmsf'])
+    _new_ihmsfs_dtype = np.dtype([(str(c), np.intc) for c in 'hmsf'])
 
     def set_jds(self, val1, val2):
         self._check_scale(self._scale)
-        if self._scale == 'utc':
+        jd1, jd2 = day_frac(val1, val2)
+        # for times before the first leap second we don't need to do anything
+        if self._scale == 'utc' and jd1 > 40000:
             # To get around leap second issues, first convert to YMD,
             # then back to astropy/ERFA-convention jd1,jd2 using the
             # ERFA dtf2d() routine which handles leap seconds.
             v1, v2 = day_frac(val1, val2)
-            (y,mo,d,f) = erfa.jd2cal(erfa.DJM0+v1,v2)
+            (y, mo, d, f) = erfa.jd2cal(erfa.DJM0+v1, v2)
             # Fractional day to HMS.  Uses 86400-second day always.
             # Seems like there should be a ERFA routine for this..
-            h = safe_kind_conversion(numpy.floor(f*24.0), dtype=int)
-            m = safe_kind_conversion(numpy.floor(numpy.remainder(f*1440.0,60)),
-                              dtype=int)
-            s = numpy.remainder(f*86400.0,60)
-            self.jd1, self.jd2 = erfa.dtf2d('UTC',y,mo,d,h,m,s)
+            # There is: erfa.d2tf. Unfortunately it takes a "number of
+            # digits" argument and returns some kind of bogus
+            # fractional-part-as-an-integer thing.
+            # Worse, it fails to provide nanosecond accuracy.
```

Comment:
@aarchiba So do you think this is a SOFA/ERFA bug?  Since I'm on the SOFA board, if you can give me a simply example of the numerical issues, it *will* get fixed.  We just did that over the last year for another small issue or two.

Reply:
I'm not totally sure whether I'm interpreting the ERFA output correctly, but if you want I can put together some tests that give astropy._erfa some significant exercise and report any bugs I come up with. Translating them back into C to be sure they're really SOFA bugs and not somewhere in the wrappers would be... more work. 

Is there documentation of the level of accuracy I should expect from various ERFA/SOFA routines, over what range of inputs?

Reply:
There is a lot of SOFA documentation.  And given that the times are stored internally with two doubles, the *intent* is to have very high precision where possible -- certainly much better than a microsecond.  I'd have to dig around to see what is expected overall.  But if there are definite instabilities at the ns level -- especially if they can be fixed easily, then those should definitely be fixed.

Reply:
Okay, well, a quick check shows that a roundtrip through d2tf and tf2d gives nanosecond accuracy for ndp=9 but does *not* give nanosecond, or even ten-nanosecond, accuracy for higher ndp. No errors or warnings are emitted. See 3cc6a8a.